algorithms/Java/strings/Boyer_Moore.java

```diff
@@ -0,0 +1,92 @@
+/* Java Program for Bad Character Heuristic of Boyer
+Moore String Matching Algorithm */
+
+
+class AWQ{
+	
+	static int NO_OF_CHARS = 256;
+	
+	//A utility function to get maximum of two integers
+	static int max (int a, int b) { return (a > b)? a: b; }
+
+	//The preprocessing function for Boyer Moore's
+	//bad character heuristic
+	static void badCharHeuristic( char []str, int size,int badchar[])
+	{
+
+	// Initialize all occurrences as -1
+	for (int i = 0; i < NO_OF_CHARS; i++)
+		badchar[i] = -1;
+
+	// Fill the actual value of last occurrence
+	// of a character (indices of table are ascii and values are index of occurence)
+	for (int i = 0; i < size; i++)
+		badchar[(int) str[i]] = i;
+	}
+
+	/* A pattern searching function that uses Bad
+	Character Heuristic of Boyer Moore Algorithm */
+	static void search( char txt[], char pat[])
+	{
+	int m = pat.length;
+	int n = txt.length;
+
+	int badchar[] = new int[NO_OF_CHARS];
+
+	/* Fill the bad character array by calling
+		the preprocessing function badCharHeuristic()
+		for given pattern */
+	badCharHeuristic(pat, m, badchar);
+
+	int s = 0; // s is shift of the pattern with
+				// respect to text
+	//there are n-m+1 potential allignments
+	while(s <= (n - m))
+	{
+		int j = m-1;
+
+		/* Keep reducing index j of pattern while
+			characters of pattern and text are
+			matching at this shift s */
+		while(j >= 0 && pat[j] == txt[s+j])
+			j--;
+
+		/* If the pattern is present at current
+			shift, then index j will become -1 after
+			the above loop */
+		if (j < 0)
+		{
+			System.out.println("Patterns occur at shift = " + s);
+
+			/* Shift the pattern so that the next
+				character in text aligns with the last
+				occurrence of it in pattern.
+				The condition s+m < n is necessary for
+				the case when pattern occurs at the end
+				of text */
+			//txt[s+m] is character after the pattern in text
+			s += (s+m < n)? m-badchar[txt[s+m]] : 1;
+
+		}
+
+		else
+			/* Shift the pattern so that the bad character
+				in text aligns with the last occurrence of
+				it in pattern. The max function is used to
+				make sure that we get a positive shift.
+				We may get a negative shift if the last
+				occurrence of bad character in pattern
+				is on the right side of the current
+				character. */
+			s += max(1, j - badchar[txt[s+j]]);
+	}
+	}
+
+	/* Driver program to test above function */
+	public static void main(String []args) {
+		
+		char txt[] = "ABAAABCD".toCharArray();

```

Comment:
Hi @Capt-Madhan , Thanks for your effort.

Please we need to check something here:
If you added this case:
```java
char txt[] = "ABCBD".toCharArray();
```
The output will be: Standard output is empty

can you please check it again!

Thanks

Reply:
Hi,
The program runs completely fine for me, Please check the attached output snapshot.
Please try in any online compiler or any other IDE.
Please let me know if the error still persists.
Thank you.

![image](https://user-images.githubusercontent.com/91775905/135742743-6452adce-e0a2-44bf-9bd8-3bef239f8742.png)

![image](https://user-images.githubusercontent.com/91775905/135742769-f93295f3-cd9f-4845-883c-2661a73c78cf.png)

Reply:
Okay @Capt-Madhan , but can you please show me this example 
```java
char txt[] = "ABCBD".toCharArray();
```

Reply:
Sure.
Example 1:
Input:     char txt[] = "ABCBD".toCharArray();
              char pat[] = "ABC".toCharArray();
Output:  Pattern found at index 0.

Example 2:
Input:     char txt[] = "ABCBD".toCharArray();
              char pat[] = "B".toCharArray();
Output:  Pattern found at index 1.
              Pattern found at index 3.
Some more Examples:
Example 3:
Input:     txt[] = "THIS IS A TEST TEXT".toCharArray();
              pat[] = "TEST".toCharArray();
Output:  Pattern found at index 10

Example 4:
Input:     txt[] =  "AABAACAADAABAABA".toCharArray();
              pat[] =  "AABA".toCharArray();
Output:  Pattern found at index 0
              Pattern found at index 9
              Pattern found at index 12

Reply:
Sure.
**Example 1:**
**Input:**    
char txt[] = "ABCBD".toCharArray();
char pat[] = "ABC".toCharArray();
**Output:**  
Pattern found at index 0.


**Example 2:
Input:**     
char txt[] = "ABCBD".toCharArray();
char pat[] = "B".toCharArray();
**Output:**  
Pattern found at index 1.
Pattern found at index 3.

**Some more Examples:**

**Example 3:
Input:**     
txt[] = "THIS IS A TEST TEXT".toCharArray();
              pat[] = "TEST".toCharArray();
**Output:**  
Pattern found at index 10


**Example 4:
Input:**    
txt[] =  "AABAACAADAABAABA".toCharArray();
              pat[] =  "AABA".toCharArray();
**Output:**  
Pattern found at index 0
              Pattern found at index 9
              Pattern found at index 12